File: 0563-binary-tree-tilt/0563-binary-tree-tilt.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function findTilt(root: TreeNode | null): number {
    return dfs(root)[1]
    
    function dfs(root: TreeNode | null): [number, number] {
        if (!root) return [0, 0]
        const [lsum, ltilt] = dfs(root.left)
        const [rsum, rtilt] = dfs(root.right)
        return [root.val + lsum + rsum, Math.abs(lsum - rsum) + ltilt + rtilt]
    }
};

File: 0563-binary-tree-tilt/0563-binary-tree-tilt.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return 0, 0
        lsum, ltilt = self.dfs(root.left)
        rsum, rtilt = self.dfs(root.right)
        return root.val + lsum + rsum, abs(lsum-rsum) + ltilt + rtilt
    
    def findTilt(self, root: Optional[TreeNode]) -> int:
        _, a = self.dfs(root)
        return a